dcd.1
.\" dcd.1 by David E. Smith. A lot of this man page was stolen from
.\" other random man pages since I don't know how to write them. :)
.\"
.\" =========================================================================
.TH DCD 1 "15 June 2002" "dcd-0.98"
.SH NAME
dcd \- play audio CDs, program CD-ROM as a smart CD player
.PD
.SH SYNOPSIS
\fBdcd\fP [\fBoptions\fP] [\fBtrack-list\fP]
.PD
.\" =========================================================================
.SH DESCRIPTION
\fIdcd\fP will help you use your CD-ROM drive as a CD player under Linux.
In addition to merely playing CDs, \fIdcd\fP can also play random tracks,
accept a pre-programmed list of tracks, and generally do most of the things
a good portable CD player is capable of.
.PD
.\" =========================================================================
.SH OPTIONS
.TP
.IP \fIdir\fP
List a directory of the CD contents (track numbers and playing times,
and track names if they are available).
This option causes \fIdcd\fP to exit immediately.
.IP \fIeject\fP
Open the CD-ROM tray, and exit immediately. \fIdcd\fP tries to locate,
and kill, other \fIdcd\fP processes that may be running.
.IP \fIhelp\fP
Display a friendly help message, and exit.
.IP \fIinfo\fP
Display one line of information about the current CD, like this:
.br
Playing track 3 (of 15), length 5:35 (of 77:05)
.br
This option causes \fIdcd\fP to exit immediately.
.IP \fIpause\fP
Pause or resume the current CD, exiting immediately.
.IP \fIstop\fP
Stop the CD currently in the player. \fIstop\fP makes a reasonable effort
to locate, and kill, other dcd processes that may be running.
.IP \fIversion\fP
Report the version of \fIdcd\fP and a brief copyright notice, and
exit immediately.
.IP \fIa\fP
Print the number of the first track of the CD to stdout, and
exit immediately. Comes in handy for various shell scripts.
.IP \fIz\fP
Print the number of the last track of the CD to stdout, and
exit immediately. Comes in handy for various shell scripts.
.IP \fIloop\fP
Loop the tracks in the track list, or the entire CD if no tracks are
specified.
An effort is made to locate, and kill, other \fIdcd\fP
processes. (Having two copies of dcd running, each with different looped
track lists, could get very strange very quickly.)
.IP \fIquit\fP
Normally, \fIdcd\fP does everything it can to get out of the way quickly,
freeing up the console or xterm it was launched from and allowing you to
get on with your day. Specifying \fIquit\fP inhibits this; i.e. \fIdcd\fP
will NOT exit until it's done. This lets you use \fIdcd\fP in things like
combined CD/MP3 playlists.
\fIdcd\fP attempts to detect unwise combinations of
commands (like loop and random, which would effectively create an infinite loop)
and nip them in the bud.
.IP \fIrandom\fP
Play randomly-selected tracks from the CD forever (until killed), and
return to the console immediately. As with \fIloop\fP, an effort is made
to deal with other rogue dcd tasks. This option may accept a tracklist.
.IP \fIx\fP
Display the CD Index discid for the CD, exiting immediately. This isn't
often useful by itself, but might be useful in a script of some sort.
.IP \fIback\fP
Go to the previous track on the CD. If you're already on the first track,
this loops around to the last track. As with many other options, we attempt
to locate and kill other instances of dcd.
.IP \fIforward\fP
Go to the next track on the CD. If you're already on the last track, this
loops around to the first track. As with many other options, we attempt to
locate and kill other instances of dcd.
.IP \fIkill\fP
Kills off other known instances of dcd. Might be useful in a shell script,
or just in case dcd somehow gets carried away and starts doing weird
things.
.\" =========================================================================
.SH EXAMPLES
.I \fIdcd\fP
Plays a CD, plain and simple.
.TP
.I \fIdcd 2\fP
Stars playing a CD from track 2.
.TP
.I \fIdcd loop\fP
Plays the whole CD, over and over and over, until killed (with dcd stop).
.TP
.I \fIdcd loop 2 4 7\fP
Plays tracks 2, 4, and 7, over and over, in that order.
.TP
.I \fIdcd random\fP
Plays random tracks from the CD, over and over and over...
.TP
.I \fIdcd random 1 3 5 7 9 12\fP
Randomly selects from the tracks listed, and plays 'em over and over...
.TP
.I \fIdcd 4 6 quit\fP
Plays tracks 4 and 6, then exits. (Normally, dcd will return you to a shell 
prompt immediately; this option is useful for some scripts, combination 
MP3/CD playlists, and other cool stuff.)

.\" =========================================================================
.SH BUGS
\fIdcd\fP doesn't make nearly as many error/sanity checks as it could.
It assumes you know what you're doing, so trying to eject a CD when the
tray is already open (as one example) could have undefined results. It
shouldn't do anything truly bad, but under the terms of the GNU General
Public License anything it somehow does isn't my fault.
.br
.br
\fIdcd\fP unfortunately now requires Internet access to work. Blame the
people at MusicBrainz for that one. When they discontinued their lean, 
sexy "CD Index" project in favor of the current project, things got a 
lot less pleasant for everyone involved (except perhaps for them). 
Fortunately, broadband access is rather more common these days, so 
hopefully this isn't as much of a problem as it would have been a couple 
years back.
.PD
.\" =========================================================================
.SH "SEE ALSO"
\fIcdcd\fP(1), \fIworkbone\fP(1)
.PD
.\" =========================================================================
.SH URL
The dcd home page is currently at:
.br
\fChttp://www.technopagan.org/dcd/\fR
.PD
.\" =========================================================================
.SH TRIVIA
`dcd' stands for Dave's CD player.
.PD
.\" =========================================================================
.SH AUTHORS
\fIdcd\fP was written by David E. Smith <dave@technopagan.org>. In
analphabetical order, the following people contributed really keen ideas
(and in some cases, actual code): Scott Walker, Luc Vrancx, Robert Tol,
Kevin Thompson, Nigel Stepp, Othmar Pasteka, Markus F.X.J. Oberhumer,
Mario Moder, Lorenz Minder, Lalo Martins, Britton Kerin, Alvaro Herrera,
Anders Semb Hermansen, Shane Henthorne, Boris Gjenero, Aidan Delaney, Izak 
Burger, Don Barber, and Jens Axboe.
.PD


ChangeLog
0.99.1 "thinner"
	Removed network support (it's in the Makefile) for Thinstation.
0.99 "sf.net"
	Basically, moved to SourceForge. That's pretty much it.
0.98 "Holy..."
     Pulled  out all the CD Index stuff, since that site apparently
       suffered heat death or something... Replaced it with massive
       ugly dependencies on libmusicbrainz.
     The above means that libstcd++, libmusicbrainz, and libm are REQUIRED.
     Added compression to local storage. Though not strictly required,
       you may wish to clean out your $HOME/.cdindex directory. (I think
       I've covered all the likely errors but I make few promises.)
     Bugs fixed after the above lobotomy: fixed the "no match found" bug.
     Added Robert Tol's "un-pause even if we're not really paused" patch.
0.96 "Nomic"
     Added Othmar Pasteka's "is there even a CD in the drive" patch which
       was so simple I shoulda done it years ago. :-)
0.95 "Not again..."
     Replaced base64.c with a more GPL-friendly version.
0.94 "Unstable Tina"
     Actually check return value for DNS lookup, since cdindex.org
       seems to be broken right about now.
     Named in honour of Unstable Tina (www.unstabletina.com), an
       incredibly cool local band. If you're in Saint Louis, drop by a show!
0.93 "One Billion"
     Fixed the "no-info" warning displayed twice bug. Not really fixed
       perhaps, but made it go away...
     Named in honour of the one billionth second of the UNIX epoch 
       (September 9, 2001, 1:46:40 AM GMT). Or Dr. Evil. Who knows?
0.92 "Lorenz"
     Named after Lorenz Minder <lminder@gmx.net> who provided a bucket
       full of useful libcdplay patches. Thanks! The most important one
       is that dcd now uses the CDROMSTART ioctl instead of CDROMSTOP.
       If this causes problems, please let me know ASAP.
     Added a compile-time option to make dcd be quiet about CDs that
       the CD Index doesn't know about.
     Truncate track names at 64 characters, so things still fit neatly
       on the standard 80-column screen.
     Plenty of file fixes, so that things like a non-existent ~/.cdindex
       directory shouldn't make dcd crash and burn anymore.
     Added "examples" to the man page.
     Reversed the changelog so newer stuff is at the top. :-)
0.90 "Quad Damage"
     Endian fix (thanks to that guy from debian)
     `z' option -- useful for shell scripts (mostly MP3 ripping)
     `a' option -- like the above but backwards. :-)
     `k' option -- kills off dcd (in case things start running away)
     I think I've finally fixed the lingering `info doesn't work' bit.
0.81 "CGB Spender" (if you don't know, don't ask)
     Bug: 0.80 doesn't like cdindex.org. Fixed.
0.80 "Felicity"
     Took out the old 'b' option which was really goofy, and added the
       new (b)ack and (f)orward options to skip about the disc. Hacks
       on dcd.c, cd-misc.c, cd-misc.h, screenop.c as appropriate.
     Finally put useful CD Index code in. cdindex.c hacked all to hell.
     Rewrote process handler code to be far more sane internally (using
       FILE instead of icky read/write). process.c beaten heavily.
     Added networking code, courtesy of LordVader @Dalnet. Still more
       rewrites on cdindex.c, screenop.c. 'connect' command option added.
0.70 "Roswell Grey"
     Added u and b options for CD Index URLs. 'x' now merely chucks out
       the disc id, which frankly isn't too useful.
     Added `dcd p' to pause/resume the disc. And it works this time too.
     libcdplay.c: do smart things when disc isn't playing (cd_current_track)
     libcdplay.c: rearranged a few boring internals
     Fixed all freeamp.org URL references to point to cdindex.org
0.60 "Feature Creep"
     Added `dcd x' to generate CD Index discids. Added base64.c and sha.c
       in support of that. (See those files for (C) info.)
     Added all PID maintenance code, thanks to Lalo Martins' nagging :)
     Re-added `dcd n' to go with PID management; it now bumps an existing
       dcd process to its next track (or next random track, or whatever)
       Added process.c which has all that stuff, including ugly filestuff.
     screenop.c: Fixed the helpscreen.
     random now can take a tracklist too.
     Rewrote changelog so it looks all professional.
     Added check for data track to libcdplay.c and screenop.c
0.51 "Syndrome"
     Added `quit' to avoid exiting immediately, works great for playlists.
     screenop.c: various cleanups, thanks to Alvaro Herrera
     Started giving versions clever names like tin prereleases.
0.50 Removed `next' and `previous' command switches. dcd.c was almost
     completely rewritten.
0.40 Added `stop CD player on SIGTERM'. Some internal cleanups. `dcd loop'
     without a tracklist now plays the whole disc.
0.35 Added `eject'


dcd.c
/* $Revision: 1.2 $ */
/*
 * This is DCD, Dave's CD Player.
 * (C)1998-2001 David E. Smith <dave@technopagan.org>
 * Released under GNU GPL. (See 'COPYING' for details.)
 *
 * This is dcd.c -- where all the fun begins.
 */

#include "libcdplay.h"
#include "screenop.h"
#include "dcd.h"
#include "infinite.h"
#include "version.h"
#include "process.h"
#include "cd-misc.h"
#include <errno.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <stdlib.h>
#include <ctype.h>
#include <getopt.h>

#define ZERO 0
#define MAX_TRACKS 103

static int tracklist[MAX_TRACKS];
/* the above should, OC, be dynamic to save memory and avoid wacky
   buffer overflows. I'll fix it later. Maybe a linked list... */
static int trk_ptr, trksz, randomflag, loopflag, quitflag;
/* trk_ptr is an index in the track array; trksz is the size of that array.
   the flags, hopefully, are obvious. */

int sigterm_handler (int sig) {
  cd_stop();
  exit(EXIT_SUCCESS);
}

int dcd_sighandler (int sig) {
  signal (sig, (void *)dcd_sighandler);
  return 0;
}

#define DCD_TOAST() { dcd_kill(SIGTERM); dcd_delpid(); } 

/* this is just ugly enough that I don't want to make it a macro */
static void dcd_check(void) {
  if(!cd_present()) {
    fprintf (stderr, "No disc in drive %s, aborting.\n", CDROM_DEVICE);
    exit(EXIT_FAILURE);
  }
}

static u_char dcd_next_track (u_char old) {
  int ret = 0;
  if (randomflag == TRUE) {
    ret = rand() % trksz;
    while (tracklist[ret] == old) ret = rand() % trksz;
    return tracklist[ret];
  } /* rand */
  trk_ptr++;
  if (loopflag == TRUE) {
    if (tracklist[trk_ptr] == ZERO) trk_ptr = 0;
    return tracklist[trk_ptr];
  } /* loop */
  else if (quitflag == TRUE) {
    if (tracklist[trk_ptr] == ZERO) return 0;
    return tracklist[trk_ptr];
  } /* quit */
  #ifdef DEBUG
    fprintf (stderr, "dcd_next_track: this shouldn't be reached\n");
  #endif
  return 0;
}

/* opt_handler and do_opt should possibly be combined, to make things
 * like x:: possible (CD Index, then pick a mode). It gets ugly though.
 */

void opt_handler (char c) {
  switch (c) {
    case 'h': print_help(); exit(EXIT_SUCCESS); break;
    case 'd': disk_directory(); exit(EXIT_SUCCESS); break;
    case 'i': print_status(); exit(EXIT_SUCCESS); break;
    case 'k': DCD_TOAST(); cd_stop();  exit(EXIT_SUCCESS); break;
    case 's': DCD_TOAST(); cd_stop();  exit(EXIT_SUCCESS); break;
    case 'e': DCD_TOAST(); cd_eject(); exit(EXIT_SUCCESS); break;
    case 'u': /* for "unpause", to make some random debian user happy */
    case 'p': cd_pause(); exit (EXIT_SUCCESS); break;
    case 'v': dcd_version(); exit(EXIT_SUCCESS); break;
    case 'r': randomflag = TRUE; break;
    case 'l': loopflag = TRUE; break;
    case 'q': 
    case 'o': quitflag = TRUE; break; /* q and o are the same */
    case 'b': DCD_TOAST(); cd_play_prev_track(); exit(EXIT_SUCCESS); break;
    case 'f': DCD_TOAST(); cd_play_next_track(); exit(EXIT_SUCCESS); break;
    case 'x': print_discid(); exit(EXIT_SUCCESS); break;
    case 'a': printf ("%i\n", cd_first_track()); exit(EXIT_SUCCESS); break;
    case 'z': printf ("%i\n", cd_last_track()); exit(EXIT_SUCCESS); break;
    default : fprintf (stderr, "Unknown option: %c\n", c);
              short_help(); exit (EXIT_FAILURE); break;
  }
}

int do_opt (int argc, char **argv) {
  int optc, i;
  int tz = cd_last_track();
  extern char *optarg;
  extern int optind;
  /* char *shortopts = "hedrilnoqvsxubfpc"; */
  char *shortopts = "bcdefhilnopqrsuvx";
  while ((optc = getopt(argc, argv, shortopts)) != EOF) opt_handler (optc);

  /* Now load a tracklist, and do a really ugly backward-compat hack.
   * This has the unintended, but maybe-useful, side effect of allowing 
   * tracks and options to be listed in any order and not requiring the
   * GNU format for options (loop instead of -loop). Ponder.
   */

  for ( ; optind < argc ; optind++)
    if (isdigit(argv[optind][0]))
       tracklist[trk_ptr++] = atoi(argv[optind]);
      else opt_handler(argv[optind][0]);
  trksz = trk_ptr;
  trk_ptr = 0;

  /* sanity check on params */

  #ifdef DEBUG
    fprintf (stderr, "flags: %s%s%s\n",
             (quitflag == TRUE) ? "quit " : "",
             (loopflag == TRUE) ? "loop " : "",
             (randomflag == TRUE) ? "random" : "");
    if ((quitflag==FALSE) && (loopflag==FALSE) && (randomflag==FALSE))
        fprintf (stderr, "none");
  #endif
  if ((quitflag == TRUE) & ((loopflag == TRUE) | (randomflag == TRUE))) {
    fprintf (stderr, "%s\n", VERSION);
    fprintf (stderr,
             "Incompatible options selected. You may not use quit and %s together.\n",
             (loopflag == TRUE) ? "loop" : "random");
    exit (EXIT_FAILURE);
  } /* if */

  #ifdef DEBUG
    fprintf (stderr, "Running sanity check on tracklist.\n");
  #endif
  for (i = 0; i < trksz; i++) {
    if (cd_data_track(tracklist[i])) {
      printf ("Track %i is a data track and can't be played. Aborting.\n", tracklist[i]);
      exit (EXIT_FAILURE);
    } /* if */
    if (tracklist[i] > tz) {
      printf ("Track %i doesn't seem to be on this disk. Aborting.\n", tracklist[i]);
      exit (EXIT_FAILURE);
    } /* if */
  } /* for */
  return trksz;
}

void init_global(void) {
  int i=0;
  if (cd_init_player(CDROM_DEVICE) == -1) {
    fprintf (stderr, "Unable to open %s, aborting.\n", CDROM_DEVICE);
    exit (EXIT_FAILURE);
  }

  randomflag = loopflag = quitflag = FALSE;
  trk_ptr = 0;
  for (i=0; i<MAX_TRACKS; i++) tracklist[i] = ZERO;
  #ifdef DEBUG
    srand(42);
  #else
    srand((unsigned int)time(NULL));
  #endif
}

int main (int argc, char **argv) {
  pid_t pid1, pid2, waitstatus;
  int i, playing;

  init_global();
  if(!cd_present()) {
    fprintf (stderr, "No disc in drive %s, aborting.\n", CDROM_DEVICE);
    exit(EXIT_FAILURE);
  }
  i = do_opt (argc, argv);

  /* "free" sanity check, rarely enabled. */
  #if DEBUG > 1
    if (i != trksz) {
      printf ("Uhoh, i != trksz, call the guru.\n");
      exit (EXIT_FAILURE);
    }
  #endif

  /* fork to have a process for actual work */
  pid1 = fork();
  switch (pid1) {
    case -1:
      perror ("unable to fork");
      exit (EXIT_FAILURE);
      break;
    case 0: break;	/* child  */
    default:
      wait (&waitstatus);
      #ifdef DEBUG
        fprintf (stderr, "parent exiting with status code %i\n", waitstatus);
      #endif
      _exit (EXIT_SUCCESS);
      break;
  } /* switch */

  /* do it again to detach from console */

  if (FALSE == quitflag) {
    pid2 = fork();
    switch (pid2) {
      case -1:
        perror ("unable to fork");
        exit (EXIT_FAILURE);
        break;
      case 0: break;
      default: exit (EXIT_SUCCESS);
    } /* switch */
  } /* if */

  /* unless quitflag set, we've detached now. */

  dcd_kill(SIGTERM);
  signal (SIGTERM, (void *)sigterm_handler);
  signal (SIGUSR1, (void *)dcd_sighandler);
  signal (SIGALRM, (void *)dcd_sighandler);
  dcd_setpid();

  /* now step through various permutations. First up: no tracklist. */

  if (tracklist[0] == ZERO) {
    if (randomflag == TRUE) shuffle_disc();   /* these will */
    else if (loopflag == TRUE) loop_disc();   /* never exit */
    else {
      cd_play_disc(cd_first_track());
      exit (EXIT_SUCCESS);
    }
  }

  /* handle the `normal' behaviour of starting from a track */

  if ((loopflag == FALSE) & (randomflag == FALSE) &
      (quitflag == FALSE) & (argc == 2)) {
     cd_play_disc(tracklist[0]);
     exit (EXIT_SUCCESS);
  } /* if */

  /* we must have a track list programmed in, let's rock. */

  #ifdef DEBUG
    fprintf (stderr, "Tracklist: ");
    for (i=0; i<MAX_TRACKS; i++)
      if (tracklist[i] != ZERO)
        fprintf (stderr, "%i  ", tracklist[i]);
    fprintf (stderr, "\n");
  #endif
  trk_ptr = -1; playing = 0; /* trust me on this. */
  while (TRUE) {
    cd_stop();
    playing = dcd_next_track(playing);
    if (playing == 0) break; /* to cover quitflag behaviour */
    cd_play_track(playing);
    #ifdef DEBUG
      fprintf (stderr, "main loop now playing track %i\n", playing);
    #endif
    alarm(cd_track_length(playing) + EXTRA_SLEEP_TIME);
    pause();
  } /* while */

  #ifdef DEBUG
    fprintf(stderr,"Exiting at the end of main()\n");
  #endif
  exit (EXIT_SUCCESS); /* This should only be reached if quitflag is set. */
  return 0; /* not reached, but makes -pedantic happy */
} /* main */
